lib/pos_leader_schedule.go

```diff
@@ -0,0 +1,106 @@
+package lib
+
+import (
+	"crypto/sha256"
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+)
+
+func (bav *UtxoView) GenerateLeaderSchedule() ([]*PKID, error) {
+	// Retrieve CurrentRandomSeedHash.
+	currentRandomSeedHash, err := bav.GetCurrentRandomSeedHash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error retrieving CurrentRandomSeedHash: ")
+	}
+
+	// Retrieve top, active validators ordered by stake.
+	validatorEntries, err := bav.GetTopActiveValidatorsByStake(int(bav.Params.LeaderScheduleMaxNumValidators))
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error retrieving top ValidatorEntries: ")
+	}
+	if len(validatorEntries) == 0 {
+		return []*PKID{}, nil
+	}
+
+	// Sum TotalStakeAmountNanos.
+	totalStakeAmountNanos := uint256.NewInt()
+	for _, validatorEntry := range validatorEntries {
+		totalStakeAmountNanos, err = SafeUint256().Add(totalStakeAmountNanos, validatorEntry.TotalStakeAmountNanos)
+		if err != nil {
+			return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error summing TotalStakeAmountNanos: ")
+		}
+	}
+
+	// Pseudocode for leader-selection algorithm:
+	// Note this is an O(N^2) algorithm where N is the number of validators we include.
+	// While len(LeaderSchedule) < len(ValidatorEntries):
+	//   Hash the CurrentRandomSeedHash and generate a new RandomUint256.
+	//   Take RandomUint256 modulo TotalStakeAmountNanos.
+	//   For each ValidatorEntry:
+	//     Skip if ValidatorPKID has already been added to the leader schedule.
+	//     If ValidatorEntry.TotalStakeAmountNanos >= RandomUint256:
+	//       Add ValidatorPKID to LeaderSchedule.
+	//       TotalStakeAmountNanos -= ValidatorEntry.TotalStakeAmountNanos.
+	//       Break out of the inner loop.
+	var leaderSchedule []*PKID
+
+	// We also track a set of ValidatorPKIDs that have already been
+	// added to the LeaderSchedule so that we can skip them when
+	// iterating over ValidatorEntries in O(1) time.
+	leaderSchedulePKIDs := NewSet([]PKID{})
+
+	for len(leaderSchedule) < len(validatorEntries) {
+		// Hash the CurrentRandomSeedHash each iteration. This generates
+		// multiple predictable pseudorandom values from the same seed.
+		currentRandomSHA256 := sha256.Sum256(currentRandomSeedHash.ToBytes())
+		currentRandomSeedHash, err = (&RandomSeedHash{}).FromBytes(currentRandomSHA256[:])
+		if err != nil {
+			return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error hashing CurrentRandomSeedHash: ")
+		}
+
+		// Take RandomUint256 % TotalStakeAmountNanos.
+		randomUint256 := uint256.NewInt().Mod(currentRandomSeedHash.ToUint256(), totalStakeAmountNanos)
```

Comment:
TODO: Make sure Mod is efficient. I'm almos certain it is.

Reply:
`uint256.Mod()` uses `uint256.QuoRem()` under the hood which looks to me like it's O(1). Link to Mod [code](https://github.com/Pilatuz/bigx/blob/v2.0.1/uint256/uint256.go#L324). Link to QuoRem [code](https://github.com/Pilatuz/bigx/blob/v2.0.1/uint256/uint256.go#L342).